CyberSource: Add error details response fields

schwarzgeist opened the pull request.

therufs commented:
hey @schwarzgeist -- this looks great! Can you put the outputs of the Cybersource remote and unit tests in a comment here? 

`ruby -Ilib:test test/remote/gateways/remote_cyber_source_test.rb` for remote tests
and 
`bundle exec rake test:units TEST=test/unit/gateways/cyber_source_test.rb` for unit tests (or the amunit/amremote shortcuts if you have them.) 

:ty:

leila-alderman commented:
This is the template that we recommend for Active Merchant commits: https://gist.github.com/jknipp/dbe9896a7fd74a0138218ae875243512.

Could you also update the commit title to `CyberSource: Add error details response fields` rather than using the Jira ticket number since this is open source?

leila-alderman commented on CHANGELOG:
```diff
@@ -32,6 +32,7 @@
 * iATS Payments: Add support for Customer Code payment method [molbrown] #3611
 * HPS: Add Google Pay support [MSmedal] #3597
 * Adyen: Parse appropriate message for 3DS2 authorization calls [britth] #3619
+* Cybersource: Add error detail [schwarzgeist] ####
```
Now that you've created a PR for this, you can fill out the number:
```suggestion
* Cybersource: Add error detail [schwarzgeist] #3629
```

schwarzgeist replied:
https://github.com/activemerchant/active_merchant/blob/63233b7e6c85f785d1636a6da0a1b42eda1ea2bd/CHANGELOG#L35

leila-alderman commented on lib/active_merchant/billing/gateways/cyber_source.rb:
```diff
@@ -941,6 +941,16 @@ def in_fraud_review?(response)
+
+      def message_from(response)
+        if response[:reasonCode] == '101' && response[:missingField]
+            response[:message] + ': ' + response[:missingField]
```
Could you use string interpolation for these instead of concatenation?
```suggestion
            "#{response[:message]}: #{response[:missingField]}"
```

schwarzgeist replied:
https://github.com/activemerchant/active_merchant/blob/63233b7e6c85f785d1636a6da0a1b42eda1ea2bd/lib/active_merchant/billing/gateways/cyber_source.rb#L947

leila-alderman commented on test/remote/gateways/remote_cyber_source_test.rb:
```diff
@@ -708,6 +708,38 @@ def test_successful_subsequent_recurring_cof_transaction
+    })
+
+    assert response = @gateway.purchase(@amount, @credit_card, @options)
+    assert !response.success?
```
There's a bit of a cleaner way to do this without the negation:
```suggestion
    assert_failure response
```

schwarzgeist replied:
https://github.com/activemerchant/active_merchant/blob/63233b7e6c85f785d1636a6da0a1b42eda1ea2bd/test/remote/gateways/remote_cyber_source_test.rb#L723

https://github.com/activemerchant/active_merchant/blob/63233b7e6c85f785d1636a6da0a1b42eda1ea2bd/test/remote/gateways/remote_cyber_source_test.rb#L739

leila-alderman commented on test/unit/gateways/cyber_source_test.rb:
```diff
@@ -851,6 +851,28 @@ def test_adds_application_id_as_partner_solution_id
     end.respond_with(successful_capture_response)
   ensure
     CyberSourceGateway.application_id = nil
+  def test_missing_field
```
It looks like you accidentally deleted an `end` statement that's necessary here:

```suggestion
  ensure
    CyberSourceGateway.application_id = nil
  end

  def test_missing_field
```

schwarzgeist replied:
Yikes

schwarzgeist replied:
https://github.com/activemerchant/active_merchant/blob/63233b7e6c85f785d1636a6da0a1b42eda1ea2bd/test/unit/gateways/cyber_source_test.rb#L853

schwarzgeist commented:
I added what seems to be a mostly correct [commit message](https://github.com/activemerchant/active_merchant/pull/3629/commits/e7dae3a9864b0983e53b97269226503a9e07868d) in one of my commits, it just didn't make it to reflect the Pull Request's details. Regardless, I'll be sure to follow the commit message's guideline's when addressing your suggestions.

schwarzgeist commented:
Actually, going to just update the Pull Request's details to avoid confusion in the meantime

schwarzgeist commented:
What should be the Pull Request's initial details

CE-193, CE-589

When an invalid/missing field is supplied, active merchant
provides  details describing exactly what field is
invalid/missing, if available for the CyberSource gateway

Unit:
78 tests, 348 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed

Remote:
77 tests, 391 assertions, 5 failures, 0 errors,
0 pendings, 0 omissions, 0 notifications
93.5065% passed

leila-alderman commented on test/unit/gateways/cyber_source_test.rb:
```diff
@@ -849,10 +849,33 @@ def test_adds_application_id_as_partner_solution_id
     end.check_request do |endpoint, data, headers|
       assert_match("<partnerSolutionID>#{partner_id}</partnerSolutionID>", data)
     end.respond_with(successful_capture_response)
-  ensure
```
Was this `ensure` line intended to be removed?

schwarzgeist replied:
Nope. Must have read it as an artifact of my "end" fix. I dug up the original version of this function and copy/pasted it back to make sure it is exactly like it was before.

https://github.com/activemerchant/active_merchant/blob/277b123d1f3da81464a6515c355635923d6a4805/test/unit/gateways/cyber_source_test.rb#L852

leila-alderman replied:
Looks good! Thanks for making those fixes!

leila-alderman commented on test/unit/gateways/cyber_source_test.rb:
```diff
@@ -849,10 +849,33 @@ def test_adds_application_id_as_partner_solution_id
+    end.check_request do |_endpoint, body, _headers|
+      assert_xml_valid_to_xsd(body)
+      assert_match %r'<c:invalidField>c:billTo/c:postalCode</c:invalidField><'m, body
+    end.respond_with(missing_field_response)
```
Should this be 
```suggestion
    end.respond_with(invalid_field_response)
```
?

schwarzgeist replied:
Yes! Good catch.

schwarzgeist replied:
https://github.com/activemerchant/active_merchant/blob/277b123d1f3da81464a6515c355635923d6a4805/test/unit/gateways/cyber_source_test.rb#L876

curiousepic commented:
@schwarzgeist FYI I'm getting failures for the two new unit tests. It also seems like these tests should be checking the response for the message, but they seem to be currently checking the request instead?
